runners/direct-java/src/main/java/org/apache/beam/runners/direct/TestStreamEvaluatorFactory.java

```diff
@@ -68,35 +83,58 @@
   @Override
   public void cleanup() throws Exception {}
 
+  /**
+   * Returns the evaluator for the provided application of {@link TestStream}, or null if it is
+   * already in use.
+   *
+   * <p>The documented behavior of {@link TestStream} requires the output of one event to travel
+   * completely through the pipeline before any additional event, so additional instances that
+   * have a separate collection of events cannot be created.
+   */
   private <InputT, OutputT> TransformEvaluator<? super InputT> createEvaluator(
       AppliedPTransform<PBegin, PCollection<OutputT>, TestStream<OutputT>> application,
       EvaluationContext evaluationContext) {
-    if (evaluator.get() == null) {
-      Evaluator<OutputT> createdEvaluator = new Evaluator<>(application, evaluationContext, inUse);
-      evaluator.compareAndSet(null, createdEvaluator);
-    }
-    if (inUse.compareAndSet(false, true)) {
-      return evaluator.get();
-    } else {
-      return null;
+    // Replace any existing value with absent, and get the existing value (atomically); ensures
+    // only one thread can obtain the evaluator per-transform.
+    Optional<Evaluator<?>> evaluator =
+        evaluators.replace(application, Optional.<Evaluator<?>>absent());
+    if (evaluator != null) {
+      return evaluator.orNull();
     }
+    Evaluator<OutputT> createdEvaluator =
+        new Evaluator<>(application, evaluationContext);
```

Comment:
And we're not worried that this means we can reuse a created evaluator outside its evaluationContext?